compiler/rustc_parse/src/parser/stmt.rs

```diff
@@ -111,6 +111,11 @@ impl<'a> Parser<'a> {
             // Do not attempt to parse an expression if we're done here.
             self.error_outer_attrs(attrs);
             self.mk_stmt(lo, StmtKind::Empty)
+        } else if self.prev_token.is_keyword(kw::Pub) && self.token.is_keyword(kw::Let) {
```

Comment:
The name of the function is `parse_stmt_without_recovery`, so I don't think it's a good idea to do recovery here. I think there was some kind of other function that gets called on invalid statements (I forgot the name but -Ztreat-err-as-bug may guide you there). Maybe you could check there?